src/plugins/HookSystem.cpp

```diff
@@ -41,31 +42,83 @@ size_t CFunctionHook::getInstructionLenAt(void* start) {
-            throw std::runtime_error("unsupported %rip usage");
-        }
-    }
-
-    return insSize;
+    return {insSize, ins};
 }
 
-size_t CFunctionHook::probeMinimumJumpSize(void* start, size_t min) {
+CFunctionHook::SInstructionProbe CFunctionHook::probeMinimumJumpSize(void* start, size_t min) {
 
-    size_t size = 0;
+    size_t              size = 0;
+
+    std::string         instrs = "";
+    std::vector<size_t> sizes;
 
     while (size <= min) {
         // find info about this instruction
-        size_t insLen = getInstructionLenAt((uint8_t*)start + size);
-        size += insLen;
+        auto probe = getInstructionLenAt((uint8_t*)start + size);
+        sizes.push_back(probe.len);
+        size += probe.len;
+        instrs += probe.assembly + "\n";
     }
 
-    return size;
+    return {size, instrs, sizes};
+}
+
+CFunctionHook::SAssembly CFunctionHook::fixInstructionProbeRIPCalls(const SInstructionProbe& probe) {
+    // analyze the code and fix what we know how to.
+    uint64_t currentAddress = (uint64_t)m_pSource;
+    // actually newline + 1
+    size_t      lastAsmNewline = 0;
+    std::string assemblyBuilder;
+    for (auto& len : probe.insSizes) {
+
+        std::string code = probe.assembly.substr(lastAsmNewline, probe.assembly.find("\n", lastAsmNewline) - lastAsmNewline);
+        if (code.contains("%rip")) {
+            CVarList       tokens{code, 0, 's'};
+            size_t         plusPresent = tokens[1][0] == '+' ? 1 : 0;
+            std::string    addr        = tokens[1].substr(plusPresent, tokens[1].find("(%rip)") - plusPresent);
+            const uint64_t OFFSET      = configStringToInt(addr);
+            if (OFFSET == 0)
+                return {};
+            const uint64_t DESTINATION = currentAddress + OFFSET + len;
+
+            if (code.starts_with("mov")) {
+                // mov +0xdeadbeef(%rip), %rax
+                assemblyBuilder += std::format("movabs $0x{:x}, {}\n", DESTINATION, tokens[2]);
+            } else if (code.starts_with("call")) {
+                // call +0xdeadbeef(%rip)
+                assemblyBuilder += std::format("pushq %rax\nmovabs $0x{:x}, %rax\ncallq *%rax\npopq %rax\n", DESTINATION);
+            } else if (code.starts_with("lea")) {
+                // lea 0xdeadbeef(%rip), %rax
+                assemblyBuilder += std::format("movabs $0x{:x}, {}\n", DESTINATION, tokens[2]);
+            } else {
+                return {};
+            }
+        } else {
+            assemblyBuilder += code + "\n";
+        }
+
+        lastAsmNewline = probe.assembly.find("\n", lastAsmNewline) + 1;
+        currentAddress += len;
+    }
+
+    std::ofstream ofs("/tmp/hypr/.hookcode.asm", std::ios::trunc);
+    ofs << assemblyBuilder;
+    ofs.close();
+    execAndGet("as /tmp/hypr/.hookcode.asm -o /tmp/hypr/.hookbinary.o && objcopy -O binary -j .text /tmp/hypr/.hookbinary.o /tmp/hypr/.hookbinary2.o");
```

Comment:
`as` may not be installed on Clang systems (e.g., FreeBSD). Try using the default compiler binary. With GCC it'd simply invoke `as` but with Clang it'd pass `-cc1as` to invoke integrated assembler.

```suggestion
    execAndGet("cc /tmp/hypr/.hookcode.asm -o /tmp/hypr/.hookbinary.o && objcopy -O binary -j .text /tmp/hypr/.hookbinary.o /tmp/hypr/.hookbinary2.o");
```

Reply:
Err, `cc ... -c ... *.as`.

Reply:
will `-c` catch that it's an asm file?

Reply:
No, GCC needs explicit `-x assembler` as it doesn't recognize `*.asm` unlike `*.s` or `*.S`. Clang works fine as is but also supports `-x assembler`.

I've edited the suggestion above.

Reply:
Alternatively, rename `.hookcode.asm` to `.hookcode.s`.

Reply:
thanks!